### DISCUSSION TITLE
Faiss Indexing unable to modify the number of centroids

### DISCUSSION THREAD

**Discussion by vrdn-23** [castorini/pyserini#1277] [Q&A] [answered]
Hi,
I've been tinkering with the Dense Indexer and was trying to index a set of documents using Faiss but I'm unable to change the number of centroids. 

This is the command I'm running and as I understand it, setting the argument should allow me to control the number of centroids that we index to.

```
(my_agent_assist) ➜  my-own-agent-assist git:(main) ✗ python -m pyserini.index.faiss --input embeddings --output faiss --hnsw --pq --efC=128 --M=128
(654, 768)
WARNING clustering 654 points to 256 centroids: please provide at least 9984 training points
WARNING clustering 654 points to 256 centroids: please provide at least 9984 training points
WARNING clustering 654 points to 256 centroids: please provide at least 9984 training points
........
WARNING clustering 654 points to 256 centroids: please provide at least 9984 training points
hnsw_add_vertices: adding 654 elements on top of 0 (preset_levels=0)
  max_level = 1
Adding 5 elements at level 1
Adding 649 elements at level 0
Done in 193.076 ms
654
```

Could someone please confirm whether this is a bug or whether I may not be handling a setting correctly?

**Comment by MXueguang** [comment, marked as answer]
Hi @vrdn-23,

You can modify it by the `--pq-nbits` option, by default it is 8, which gives 2^8 = 256 centroids. https://github.com/castorini/pyserini/blob/11124e9bc5288e7ebca22d2759a292cf507afe2a/pyserini/index/faiss.py#L72

**Comment by vicpara** [comment]
If I'm using HSNW and PQ it seems that changing the number of centroids is impossible at the time. Is there any reason for it?

https://github.com/castorini/pyserini/blob/1f6ba7754bb9f8656232658e4e1cab8b299b7505/pyserini/index/faiss.py#L64

**Reply by lintool** [reply]
No reason, just haven't connected the Faiss settings to cmdline params. PRs welcome!
